### PR TITLE
Generate Youtube timestamps

### PR DESCRIPTION
**Pull request by moorecp**
This is sort of a discussion on this proof of concept feature. The current implementation works by displaying a button under the Edit Video input on the edit run page if the video link is a Youtube video. This displays a modal where the user can input an offset in seconds for when the first split starts in the video. This populates a textarea with the timestamps to copy/paste into your Youtube description.

image

image

Things to discuss:

- Should the mechanism for generating the timestamps always be available even if you don't have a Youtube video attached to the run?
- Do the timestamps need to be displayed, or should there just be a button to copy the timestamps?
- Should the offset be specified for the timestamps? Currently, when syncing the video with the splits in the timeline, it simply adds 10 seconds. Should there instead be a way to specify the offset more generally such that it also applies to the timeline (or defaults to the current 10 seconds if not provided)?

### CODE REVIEW

**Comment by glacials** [comment]
This is awesome! 😍 

> Should the mechanism for generating the timestamps always be available even if you don't have a Youtube video attached to the run?

I think we should hide it until you tell us you have a YouTube video, but can we use JavaScript to show it instantly when you paste your YouTube link into the field (before you hit save)? I'm afraid that if we don't show it until they hit Save, their eyes will already be wandering and they'll miss it.

> Do the timestamps need to be displayed, or should there just be a button to copy the timestamps?

Displaying them is a nice way to make the feature's helpfulness "click" right away, but I think we can do it more succinctly. Maybe use a tooltip to display a short description and a 2-3 line truncated version?

> Should the offset be specified for the timestamps? Currently, when syncing the video with the splits in the timeline, it simply adds 10 seconds. Should there instead be a way to specify the offset more generally such that it also applies to the timeline (or defaults to the current 10 seconds if not provided)?

That's brilliant! I am 100% on board with saving the offset more generally and using it to make the run page embedded video more accurate. Thanks to your code last year splitting off the `Video` model, this should be a super clean addition.

So with all this in mind, what if we change this video card to ask for two fields now, URL and offset (for any video type), and if a YouTube URL is pasted in the URL field we use JavaScript to show a "Copy YouTube timestamps" button with a tooltip explaining it?

**Comment by moorecp** [comment]
> So with all this in mind, what if we change this video card to ask for two fields now, URL and offset (for any video type), and if a YouTube URL is pasted in the URL field we use JavaScript to show a "Copy YouTube timestamps" button with a tooltip explaining it?

I think that's the best approach. I'll make those changes.

Do you want the changes to the timeline splits/video sync on this branch or a separate one?

**Comment by glacials** [comment]
I don't have a preference either way, whatever is best for you.

File: app/assets/stylesheets/brands.sass [hunk 1]
```diff
@@ -10,6 +10,8 @@ $patreon-faded: rgba(230, 70, 26, 0.8)
 $srdc:       rgba(117, 102, 76, 1)
 $srdc-faded: rgba(117, 102, 76, 0.8)
 
+$youtube:       rgba(255, 0, 0, 1)
+$youtube-faded: rgba(255, 0, 0, 0.8)
```
[the remark this answers is not in the document]
Comment (moorecp) [reply app/assets/stylesheets/brands.sass:14]: Right? I had to check twice to be sure.

File: db/schema.rb [hunk 2]
```diff
@@ -516,6 +516,7 @@
     t.string "url", null: false
     t.datetime "created_at", precision: 6, null: false
     t.datetime "updated_at", precision: 6, null: false
+    t.integer "start_offset", default: 10
```
[the remark this answers is not in the document]
Comment (moorecp) [reply db/schema.rb]: Makes sense.

File: db/migrate/20200705172000_backfill_add_start_offset_to_videos.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class BackfillAddStartOffsetToVideos < ActiveRecord::Migration[6.0]
+  disable_ddl_transaction!
+
+  def change
+    Video.find_in_batches do |records|
+      Video.where(id: records.map(&:id)).update_all start_offset: 10
+    end
```
[the remark this answers is not in the document]
Comment (moorecp) [reply db/migrate/20200705172000_backfill_add_start_offset_to_videos.rb]: 👍

File: app/javascript/youtube_timestamps.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    const segmentStart = moment.duration(duration).format(format, {trim: false});
+    return `${segmentStart} ${htmlDecode(segment.name)}`
+  })
+  if (offsetInMilliseconds > 0) {
+    offsets.unshift("0:00 Start of Video");
+  }
+  textarea.value = offsets.join("\n")
+  textarea.select();
+  document.execCommand("copy");
+}
```
[the remark this answers is not in the document]
Comment (moorecp) [reply app/javascript/youtube_timestamps.js]: Yup.

File: db/migrate/20200705172000_backfill_add_start_offset_ms_to_videos.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
+class BackfillAddStartOffsetMsToVideos < ActiveRecord::Migration[6.0]
+  disable_ddl_transaction!
+
+  def change
+    Video.find_in_batches do |records|
+      Video.where(id: records.map(&:id)).update_all start_offset_ms: 10_000
```
[the remark this answers is not in the document]
Comment (moorecp) [reply db/migrate/20200705172000_backfill_add_start_offset_ms_to_videos.rb:6]: I assume so. This was just copy/pasted from another `strong_migrations` migration that might be older and doing things a different way for some reason. Looking at the [docs](https://github.com/ankane/strong_migrations#good-2) for `strong_migrations`, it does just do `update_all`. Although, it also has a 0.01 second sleep in to throttle the queries, as well. So, I guess I'll add that, too.

File: db/migrate/20200705172000_backfill_add_start_offset_ms_to_videos.rb [hunk 6]
```diff
@@ -0,0 +1,12 @@
+class BackfillAddStartOffsetMsToVideos < ActiveRecord::Migration[6.0]
+  disable_ddl_transaction!
+
+  def change
+    Video.find_in_batches do |records|
```
[the remark this answers is not in the document]
Comment (moorecp) [reply db/migrate/20200705172000_backfill_add_start_offset_ms_to_videos.rb:5]: Better safe than sorry.

File: app/javascript/youtube_timestamps.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  if (event.target.id !== 'video_url') {
+    return
+  }
+
+  const visibility = RegExp(/^(?:https?:\/\/)?(?:www\.)?youtu(?:be\.com|\.be)\/.+/i).test(event.target.value) ? 'visible' : 'hidden'
+  document.getElementById('youtube-timestamps-button').style.visibility = visibility
+})
+
+document.addEventListener('click', event => {
+  if (event.target.id !== 'youtube-timestamps-button') {
```
[the remark this answers is not in the document]
Comment (moorecp) [reply app/javascript/youtube_timestamps.js:11]: Thanks.

File: app/views/runs/edit.slim [hunk 8]
```diff
@@ -29,7 +29,22 @@
...
+          = f.text_field :url, class: 'form-control', placeholder: 'URL for a Twitch or YouTube VOD', 'aria-label' => 'Video URL'
+        = f.label :start_offset_seconds, "Start Offset (in seconds)"
+        div
+          = f.number_field :start_offset_seconds, class: 'form-control', 'aria-label' => 'Video Start Offset'
+        a.btn.btn-youtube.mt-2#youtube-timestamps-button(
+          content='Click here to copy timestamps of your splits for pasting into your Youtube description'
+          href='#'
+          v-tippy=true
+          style="visibility: #{@run.belongs_to?(current_user) && @run.video&.youtube? ? 'visible' : 'hidden'};"
+          onclick="event.preventDefault(); window.copyYoutubeTimestamps(); return false;"
```
[the remark this answers is not in the document]
Comment (moorecp) [reply app/views/runs/edit.slim:43]: Whoops, yup.

File: app/views/runs/edit.slim [hunk 9]
```diff
@@ -29,7 +29,22 @@
...
+        )
+          => icon('fab', 'youtube')
+          ' Copy Youtube Timestamps
```
[the remark this answers is not in the document]
Comment (moorecp) [reply app/views/runs/edit.slim:46]: Indeed it does.

**Comment by moorecp** [comment]
> I also had one bug appear on me, if I copy timestamps for [this run](https://splits.io/gcb) the last split is marked as `0:10` even though it starts over 1 hour into the run. Full timestamps I got were:

Turns out this is also a bug in the segment timeline with navigating for videos. This segment is a `collapsed_segment` with 3 segments, 2 of which are skipped. So, if you look at the `start_ms` for that segment, it's 0ms. I can fix this for just this feature, but I have an idea that could help for all situations where we just want a start time for a `collapsed_segment`.

**Comment by glacials** [comment]
Oh, nice find! Yeah if `start_ms` for collapsed segments are all 0 then we definitely want to fix that universally, but no need to do so in this branch. Everything looks good to me 👍

**Pull request merged**